packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1347,45 +1327,88 @@ module.exports = function<T, P, I, TI, PI, C, CC, CX, PL>(
-      // these magic numbers entirely by always comparing to the current time?
-      if (expirationTime === Sync) {
-        if (isUnbatchingUpdates) {
-          performWork(Sync, null);
-        } else {
-          performWork(Task, null);
-        }
-      } else if (!isCallbackScheduled) {
-        isCallbackScheduled = true;
-        scheduleDeferredCallback(flushAsyncWork);
+  function removeFromSchedule(previousRoot: FiberRoot, root: FiberRoot) {
+    if (root === root.nextScheduledRoot) {
+      root.nextScheduledRoot = null;
+      firstScheduledRoot = lastScheduledRoot = null;
+      return;
+    }
+    if (root === firstScheduledRoot) {
+      const next = root.nextScheduledRoot;
+      firstScheduledRoot = next;
+      lastScheduledRoot.nextScheduledRoot = next;
+      root.nextScheduledRoot = null;
+      return;
+    }
+    if (root === lastScheduledRoot) {
+      lastScheduledRoot = previousRoot;
+      lastScheduledRoot.nextScheduledRoot = firstScheduledRoot;
+      root.nextScheduledRoot = null;
+      return;
+    }
+    previousRoot.nextScheduledRoot = root.nextScheduledRoot;
+    root.nextScheduledRoot = null;
+  }
+
+  // requestWork is called by the scheduler whenever a root receives an update.
+  // It's up to the renderer to call renderRoot at some point in the future.
+  function requestWork(root: FiberRoot, expirationTime: ExpirationTime) {
+    if (nestedUpdateCount > NESTED_UPDATE_LIMIT) {
+      invariant(
+        false,
+        'Maximum update depth exceeded. This can happen when a ' +
+          'component repeatedly calls setState inside componentWillUpdate or ' +
+          'componentDidUpdate. React limits the number of nested updates to ' +
+          'prevent infinite loops.',
+      );
+    }
+
+    // Add the root to the schedule.
+    addToSchedule(root, expirationTime);
+
+    if (isRendering) {
+      // Prevent reentrancy. Remaining work will be scheduled at the end of
+      // the currently rendering batch.
+      return;
+    }
+
+    if (isBatchingUpdates) {
+      // Flush work at the end of the batch.
+      if (isUnbatchingUpdates) {
+        // ...unless we're inside unbatchedUpdates, in which case we should
+        // flush it now.
+        flushRoot(root, Sync);
       }
+      return;
+    }
+
+    // TODO: Get rid of Sync and use current time?
+    if (expirationTime === Sync) {
+      performWork(Sync, null);
+    } else if (!isCallbackScheduled) {
+      isCallbackScheduled = true;
+      scheduleDeferredCallback(flushAsyncWork);
     }
   }
 
   function findHighestPriorityRoot() {
     let highestPriorityWork = NoWork;
     let highestPriorityRoot = null;
 
-    let previousScheduledRoot = null;
+    let previousScheduledRoot = lastScheduledRoot;
     let root = firstScheduledRoot;
-    while (root !== null) {
+    let shouldContinue = true;
+    while (root !== null && shouldContinue) {
+      if (root === lastScheduledRoot) {
+        shouldContinue = false;
+      }
```

Comment:
Since I'm pretty sure you'll comment on this, I think this could be improved by using if statement + a do-while loop.

Reply:
Eh nah it's better this way.

Reply:
Never mind, got it to work with just control flow.

Reply:
I was just reloading to comment. :) Thanks!

Reply:
Now go watch some Netflix. You've earned it!

Reply:
For the record, I wasn't working on this for the whole two hours. I ate dinner then read a book and then came back to it :D